docs/proposals/ingress-l4.md

```diff
@@ -0,0 +1,346 @@
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Ingress L4 Proposal
+
+## Abstract
+
+A proposal to add support for TCP and UDP to the Ingress Resource.
+
+Existing issue regarding Ingress and TCP
+* Ingress and TCP [#23291](https://github.com/kubernetes/kubernetes/issues/23291)
+
+
+## Use Cases
+
+1. Be able to accept and route traffic for arbitrary TCP ports
+1. Be able to accept and route traffic for arbitrary UDP ports
+
+
+## Motivation
+
+When deploying and exposing non-http(s) based services, Kubernetes users are
+limited to NodePort and service type=LoadBalancer as methods for exposing these
+services using interfaces supported by the Kubernetes API.
+
+Other methods exist, via contributed code, for exposing non-http(s) based services:
+
+* [keepalived-vip](https://github.com/kubernetes/contrib/tree/master/keepalived-vip)
+* NGINX Ingress Controller
+  [nginx-ingress](https://github.com/kubernetes/contrib/tree/master/ingress/controllers/nginx)
+
+However these implementations rely on a ConfigMap being populated with a
+mapping of destination endpoint to service.  A single common interface for
+describing all Ingress would is highly desirable.
+
+
+## Design Considerations
+
+### State of L4 Protocol Support in Cloud Providers
+
+L4 support is not uniform across all Cloud Providers, specifically TLS
+termination is lacking in several Cloud Providers.
+
+|           | TCP | UDP | TLS |
+| --------- | --- | --- | --- |
+| AWS       |  x  |     |  x  |
+| GCE       |  x  |  x  | (1) |
+| Mesos     |  ?  |  ?  |  ?  |
+| Openstack |  x  |  x  |  x  |
+| Rackspace |  x  |  x  |     |
+| vSphere   |  x  |  x  |  x  |
+
+(1) GCE has TLS as an alpha service
+
+### TCP and UDP support
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - tcp:
+        port: 53
+      backend:
+        serviceName: testdns
+        servicePort: 53
+    - udp:
+        port: 53
+      backend:
+        serviceName: testdns
+        servicePort: 53
+```
+
+While HTTP rules have host specifiers, in TCP and UDP rules they would not be
+allowed as hostname handling is specific to L5 or higher layer protocol being
+used over TCP or UDP.  TCP and UDP as protocol specifiers should be agnostic to
+higher layer protocols.
+
+### Ingress type specifier
+
+The current Ingress API assumes a protocol based on the presence or absence
+of a key (tls, http).  This leaves open the possibility of a spec that defines
+all protocols.
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - tcp:
+        port: 3306
+      udp:
+        port: 3306
+      host: www.foo.com
+      tls:
```

Comment:
While true that TLS is a security layer, to the best of my recollection it's not actually carrying TCP traffic as much as it's carrying an encrypted version of the TCP payload.  There's no encapsulated protocol header like with IPSec, so the plaintext version of the application data carried by TLS is expected to be consumed by a TCP service, not UDP.  So I'm not sure it makes sense to allow UDP rules to be specified along with a TLS section in the spec.  For me this comes back to clarity of the grammar, if a security feature were available and expressable in a consistent way for all supported protocols/rules, I think it makes sense to allow the security parameters to be specified at a level above the specific rules.  However, since this is not true I don't think it's best to have security at a level above the individual rules.

That all said, I also agree about keeping things simple and would like to separate the discussion of how best to represent TLS separately from that of "should we add TCP as a rule?"